_posts/2017-09-24-Unit-testing-in-angular.md

```diff
@@ -0,0 +1,504 @@
+        const component = new TimeoutExample();
+        expect(component.counter).toBe(0);
+        component.updateCounterWithDelay();
+        tick();
+        expect(component.counter).toBe(0);
+        tick(10);
+        expect(component.counter).toBe(0);
+        tick(90);
+        expect(component.counter).toBe(1);
+    }));
+{% endhighlight %}  
+
+Although it isn't really a useful test, it clearly shows how the tick function manipulates the advancement of time.
+
+### Observables
+
+Now, what if you want to test a function that returns an Observable? 
+Well, simply subscribe to it in an async block and check the result!
+
+{% highlight coffeescript %}  
+    it('should return a list of cars' async(() => {
+        service.findAll().take(1).subscribe(
+            (result) => {
+                expect(result.length).toBe(9);
+            },
+            (error) => {
+                expect(true).toBeFalsy();
+            }
+        );
+    }));    
+{% endhighlight %}  
+
+The error clause may seem strange. 
+However, what if the `findAll()` call returns an error and you don't have the error clause in your test? 
+You'll simply think that your test has passed because it appears green in the console. 
+With code coverage enabled, you may notice that the part of the code you were testing isn't marked as covered. 
+By adding `expect(true).toBeFalsy();` to the error clause, your test will fail because it shouldn't get there!
+
+# What to test
+Now that we know a little on how to test, let's have a look at what to test. 
+
+For starters, you don't have access to private and protected variables/functions, so all you can do is test the public ones. 
+All variables that are accessed by the view, should be public so those are the ones you can already test. 
+The constructor and all lifecycle events can be called too as they are public. 
+You should never ever set a variable or function to public in order to test it. 
+If you can't test it because it's private, you're doing something wrong. 
+You should be able to get to it through other functions.
+
+Generally, you give an input and assert the output, it's as simple as that. 
+Unit tests in Karma also allow you to assert whether a function has been called and optionally with which parameters (`toHaveBeenCalled` and `toHaveBeenCalledWith`). 
+This can be usefull when for example testing a `void` function that calls a mocked function. 
+That way you can still assert the output. 
+So, think of possible scenarios for the functions to test, provide the input and assert the output. 
+Also try to cover other paths than just the happy paths.
+
+Testing getters and setters usually isn't needed, unless they are more complex. 
+In most cases they're not and it's quite pointless to call a setter and then assert whether it has been set correclty. 
+Most of the time, these will be called indirectly when testing other functions.
+
+Also, I don't see the point in testing functions like this:
```

Comment:
imo you should definitely test these implementations, since you want to verify the correct endpoint is being called.

f.e. a new api version is being release with breaking changes and you need to change the path in the url from v1 to v2

Reply:
So you test that http.get got the parameters you passed it?

Reply:
yep could do.

I use MockBackend, f. e. : 
```
  let service: MyHttpService;
  let injector: Injector;
  let connection: MockConnection;
  let backend: MockBackend;

  beforeEach(() => {
    injector = ReflectiveInjector.resolveAndCreate([
      MyHttpService,
      Http,
      {provide: ConnectionBackend, useClass: MockBackend},
      {provide: RequestOptions, useClass: BaseRequestOptions},
      {provide: ENVIRONMENT_CONFIG_TOKEN, useValue: environmentConfigDevStub}
    ]);

    service = injector.get(MyHttpService);
    backend = injector.get(ConnectionBackend);
    environment = injector.get(ENVIRONMENT_CONFIG_TOKEN);
  });

  describe('when POST resource', () => {
    beforeEach(() => {
      service.post(resource).subscribe();
      connection = backend.connectionsArray[0];
    });
    it('should be a connection', () => {
      expect(connection).toBeDefined();
    });
    it('should make a post', () => {
      expect(connection.request.method).toEqual(RequestMethod.Post);
    });
    it('should make a request to v1/resourceEndpoint', () => {
      expect(connection.request.url).toEqual(`${environment.baseUrlGatewayAPI}v1/...`);
    });
  });
```

Reply:
I think that's a little overkill to go and test all http calls like that. It seems like you're testing the angular http library itself. I simply expect http.get (or post, patch...) to return an observable of a Response (or an object with the new HttpClient).

But, I'll leave out the part about not having to test it :)

Reply:
yeah it's pretty verbose :), but I believe this is a good way to test, especially if you have specific headers, multiple query parameters, etc.

also im not rly testing the angular http lib, im testing the implementation of my httpservice post method **service.post(resource);**  and I expect it to be using the angular http client library (mockbackend).

mockbackend also comes with various utiliy api's like mockRespond, mockError.

pure unit test I would spyOn http.get, and verify it has been called with the correct parameters just like you said.